### PR TITLE
test(all): switch to use GitHub strategy matrix and fix flaky tests

### PR DESCRIPTION
**Pull request by ijemmy**
## Description of your changes

The e2e tests have been flaky due to following reasons:
1. Some tests (mostly Logger) time out. **Cause:** Refactoring tracing increase the number of concurrent deployment. This slows down the overall run and we are more likely to hit timeout. (GitHub runner has only 2 cores)
2. Tracer tests sometimes fail from not having a particular field (`error`). **Causes:** We removed the explicit wait on Tracer during refactoring. We now rely on polling until we receive the expected number of traces. However, not all subsegments are available in the traces yet. 

This PR addresses the issues by:
1. Increases timeout
2. Use matrix strategies to run 6 combinations in 6 different runners ( 2 Lambda runtimes x 3 packages)
3. Add a logic on polling to retry if the subsegments are not available yet.

### How to verify this change
You can trigger the e2e test workflow manually.

I've run this 5 times without any failure so far. > https://github.com/awslabs/aws-lambda-powertools-typescript/actions/runs/2295223762. 

### Related issues, RFCs
#825  <-- the log will be descriptive with Matrix strategy

### PR status

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

## Checklist

- [x] [My changes meet the tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [x] I have made corresponding changes to the *documentation*
- [x] I have made corresponding changes to the *examples*
- [x] My changes generate *no new warnings*
- [x] The *code coverage* hasn't decreased
- [x] I have *added tests* that prove my change is effective and works
- [x] New and existing *unit tests pass* locally and in Github Actions
- [x] Any *dependent changes have been merged and published* in downstream module
- [x] The PR title follows the [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

- [ ] I have documented the migration process
- [ ] I have added, implemented necessary warnings (if it can live side by side)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/commons/tests/utils/e2eUtils.ts [hunk 1]
```diff
@@ -57,7 +57,7 @@ export const createStackWithLambdaFunction = (params: StackWithLambdaFunctionOpt
 };
 
 export const generateUniqueName = (name_prefix: string, uuid: string, runtime: string, testName: string): string => 
-  `${name_prefix}-${runtime}-${testName}-${uuid}`.substring(0, 64);
+  `${name_prefix}-${runtime}-${uuid.substring(0,5)}-${testName}`.substring(0, 64);
```
Comment (ijemmy) [inline packages/commons/tests/utils/e2eUtils.ts:60]: When test name is really long, the uuid is truncated on some resources. This results in name clashing when the same test cases are run at the same time.

**Review by dreamorosi — DISMISSED** [review]
Already looks good and passing!

As only note I'd recommend removing `concurrently` from the `package-lock.json`. Since we are no longer using it we can avoid having to maintain future updates for it.

File: .github/workflows/run-e2e-tests.yml [hunk 2]
```diff
@@ -2,48 +2,63 @@ name: run-e2e-tests
...
+          npm run test
+  e2e-tests:
     runs-on: ubuntu-latest
     permissions:
       id-token: write # needed to interact with GitHub's OIDC Token endpoint.
       contents: read
+    strategy:
+      matrix:
+        version: [12, 14]
+        package: [logger, metrics, tracing]
```
Comment (saragerion) [inline .github/workflows/run-e2e-tests.yml:42]: Outside of the scope of this PR, but we should rename the tracer folder to "tracer" for consistency. Created an issue:
https://github.com/awslabs/aws-lambda-powertools-typescript/issues/829

File: packages/logger/tests/e2e/constants.ts [hunk 3]
```diff
@@ -1,5 +1,6 @@
 export const RESOURCE_NAME_PREFIX = 'Logger-E2E';
-export const TEST_CASE_TIMEOUT = 20_000; // 20 seconds
-export const SETUP_TIMEOUT = 300_000; // 300 seconds
-export const TEARDOWN_TIMEOUT = 200_000; 
+export const ONE_MINUTE = 60 * 1000;
+export const TEST_CASE_TIMEOUT = ONE_MINUTE;
+export const SETUP_TIMEOUT = 5 * ONE_MINUTE;
+export const TEARDOWN_TIMEOUT = 5 * ONE_MINUTE;
```
Comment (saragerion) [inline packages/logger/tests/e2e/constants.ts:5]: I like this syntax. More readable!

File: packages/tracing/tests/helpers/tracesUtils.ts [hunk 4]
```diff
@@ -14,11 +14,16 @@ import {
   expectedCustomResponseValue, 
   expectedCustomErrorMessage,
 } from '../e2e/constants';
+import { FunctionSegmentNotDefinedError } from './FunctionSegmentNotDefinedError';
```
Comment (saragerion) [inline packages/tracing/tests/helpers/tracesUtils.ts:17]: Question: what do you mean here with segment not defined?

File: packages/tracing/tests/helpers/tracesUtils.ts [hunk 5]
```diff
@@ -14,11 +14,16 @@ import {
...
 interface ParsedDocument {
   name: string
   id: string
   start_time: number
   end_time: number
+  // This flag may be set if the segment hasn't been fully processed
+  // The trace may have already appeared in the `getTraceSummaries` response 
+  // but a segment may still be in_progress
+  in_progress?: boolean
```
Comment (saragerion) [inline packages/tracing/tests/helpers/tracesUtils.ts:26]: By looking at the documentation, also the `end_time` key might be not defined, correct? 

> end_time – number that is the time the segment was closed. For example, 1480615200.090 or 1.480615200090E9. Specify either an end_time or in_progress.

https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-fields

File: packages/tracing/tests/helpers/tracesUtils.ts [hunk 6]
```diff
@@ -117,6 +122,28 @@ const getTraces = async (xrayClient: XRay, startTime: Date, resourceArn: string,
...
+        invocationSubsegment = getInvocationSubsegment(trace);
+      } catch (error) {
+        if (error instanceof FunctionSegmentNotDefinedError){
+          retry(new Error(`There is no Function subsegment (AWS::Lambda::Function) yet. Retry.`));
+        } else {
+          throw error;
+        }
+      }
+
+      retryFlag = retryFlag || (!!invocationSubsegment.in_progress);
```
Comment (saragerion) [inline packages/tracing/tests/helpers/tracesUtils.ts:141]: Question: essentially here we are looking at the `in_progress` key. If it's set, we retry. Correct?

**Review by saragerion — COMMENTED** [review]
Thanks for the PR! I left some comments/questions but already looking good :-)

**Review by dreamorosi — APPROVED** [review]

**Review by saragerion — APPROVED** [review]

File: packages/tracing/tests/helpers/tracesUtils.ts (hunk 4 above)
Comment (ijemmy) [reply packages/tracing/tests/helpers/tracesUtils.ts:17]: If X-Ray hasn't fully processed all segments, some of them may be missing.
In our case, we may try to get the Function (AWS::Lambda::Function) segment too early. I use this custom error to distinguish the type of error form others.

File: packages/tracing/tests/helpers/tracesUtils.ts (hunk 5 above)
Comment (ijemmy) [reply packages/tracing/tests/helpers/tracesUtils.ts:26]: Yes. In my experiments, `end_time` field doesn't exist.  It should be optional.

image

File: packages/tracing/tests/helpers/tracesUtils.ts (hunk 6 above)
Comment (saragerion) [reply packages/tracing/tests/helpers/tracesUtils.ts:141]: I see! thanks for clarifying

File: .github/workflows/run-e2e-tests.yml (hunk 2 above)
Comment (dreamorosi) [reply .github/workflows/run-e2e-tests.yml:42]: #854

**Pull request merged**